ruma-events: Add secret sharing.

Frinksy opened the pull request:
This is a first PR working towards https://github.com/ruma/ruma/issues/478. It adds the [Secret Sharing](https://github.com/matrix-org/matrix-doc/blob/master/proposals/1946-secure_server-side_storage.md#sharing) part of MSC 1946.

Two questions I have:
- what's the policy on adding constructors for events? I see some have them and some don't.
- would it be useful to write tests for these two events, or are they small enough that it doesnt matter?

I also used the `String` type for secrets, instead of `ruma_identifiers::ClientSecret` as neither the unstable spec nor the MSC text specify a grammar for the secrets.

Frinksy committed:
ruma-events: Add secret sharing.

jplatte commented:
All `non_exhaustive` structs need constructor functions, otherwise they can't be created. Where are you seeing event content types that don't have constructors?

Frinksy commented:
I hesitated to add constructors because I saw `ChildEventContent` and `ParentEventContent` which didn't have constructors and are `non_exhaustive`.
I'll add constructors for the two structs I added.

poljar commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,52 @@
+    /// The name of the secret that is being requested.
+    ///
+    /// Required if `action` is `request`.
+    pub name: Option<String>,
```
I think it would make sense for this to be part of the `RequestAction` enum so we encode that the action of type `request` requires a name.

We might also want to consider using an enum for this instead there are some known names that are used, e.g. for the private cross signing keys:
> The cross-signing private keys can be stored on the server or shared with other devices using the Secrets module. When doing so, the master, user-signing, and self-signing keys are identified using the names m.cross_signing.master, m.cross_signing.user_signing, and m.cross_signing.self_signing

For the backup key:

> The recovery key can also be stored on the server or shared with other devices using the Secrets module. When doing so, it is identified using the name m.megolm_backup.v1, and the key is base64-encoded before being encrypted.

Frinksy replied:
I tried applying your suggestions with the new `RequestAction` and `SecretName` types. Let me know if it's in line with what you expect, or if there are other known names that I missed and could add.

poljar reviewed (commented):
Oh neat, I'll need that soon.

> would it be useful to write tests for these two events, or are they small enough that it doesn't matter?

I would say yes please, Ruma usually does some transforms to represent the deserialized data a bit neater, those transforms can break stuff in non-obvious ways.

jplatte commented:
> I hesitated to add constructors because I saw `ChildEventContent` and `ParentEventContent` which didn't have constructors and are `non_exhaustive`.

I see. That's a bug, if you want you can fix it yourself in a separate PR, otherwise could you open an issue about it?

Frinksy committed:
Add constructors for SecretSendEventContent and SecretRequestEventContent.

Frinksy committed:
Include name in `RequestAction` and add tests.

Frinksy committed:
Add CHANGELOG entry for secret sharing.

Frinksy committed:
Add `SecretName` enum for m.secret.request.

poljar commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,165 @@
+    CrossSigningSelfSigningKey,
+
+    /// Recovery key (m.megolm_backup.v1.recovery_key).
+    #[ruma_enum(rename = "m.megolm_backup.v1.recovery_key")]
```
Is this right? I can't find a mention of `m.megolm_backup.v1.recovery_key` in the spec, as far as I can tell the name is just `m.megolm_backup.v1` mentioned over [here](https://spec.matrix.org/unstable/client-server-api/#recovery-key):

> The recovery key can also be stored on the server or shared with other devices using the Secrets module. When doing so, it is identified using the name m.megolm_backup.v1, and the key is base64-encoded before being encrypted.

Frinksy replied:
For the key names I had only based myself off of the MSC text [here](https://github.com/uhoreg/matrix-doc/blob/ssss/proposals/1946-secure_server-side_storage.md#secret-storage):

> Encrypted data is stored in the user's account_data using the event type defined by the feature that uses the data. For example, decryption keys for key backups could be stored under the type `m.megolm_backup.v1.recovery_key`, or the self-signing key for cross-signing could be stored under the type `m.cross_signing.self_signing`.

It does seem that it should be `m.megolm_backup.v1`, since that's what's used in the more recent MSCs, the unstable spec and matrix-js-sdk.

poljar replied:
Right, let's go with what's in the unstable spec and in the js-sdk.

poljar commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,165 @@
+    use serde_json::{json, to_value as to_json_value};
+
+    #[test]
+    fn secret_request_serialization() {
```
Can we test deserialization as well? It's not uncommon for only one of them to be broken.

Frinksy replied:
I've added tests, and it turns out the derived `Deserialization` implementation didn't work.
I wasn't sure about what to do with a custom `action` variants when deserializing, so I made it discard the `name` field the same way it does when the `action` is `"request_cancellation"`.

jplatte replied:
I think discarding the secret name for custom actions makes sense, and it's forwards compatible with a richer API for custom actions later.

poljar reviewed (commented):
This looks much nicer, though one of the names seems to be incorrect and a deserialization test would be nice as well.

poljar reviewed (commented):
Just realized, this is missing entries for the event and content enums.

jplatte commented:
> Just realized, this is missing entries for the event and content enums.

You mean in [enums.rs](https://github.com/ruma/ruma/blob/main/crates/ruma-events/src/enums.rs), right?

poljar commented:
> > Just realized, this is missing entries for the event and content enums.
> 
> You mean in [enums.rs](https://github.com/ruma/ruma/blob/main/crates/ruma-events/src/enums.rs), right?

Yeah.

Frinksy committed:
Rename recovery key and add Deserialize impl for `SecretRequestEventContent`.

Rename to "m.megolm_backup.v1".
Add manual deserialization for `SecretRequestEventContent`.
Add deserialization tests.

Frinksy committed:
Add m.secret.send and m.secret.request to enums.rs.

poljar commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,381 @@
+}
+
+/// The name of a secret.
+#[derive(Clone, Debug, StringEnum)]
```
Why doesn't this enum have all the goodness the `EventEncryptionAlgorithm` for example has? `PartialEq` and `AsRefStr` would be very useful for this.

Frinksy replied:
I've added `PartialEq, Eq, PartialOrd, Ord` to both `SecretName` and `RequestAction`.
`AsRefStr, DisplayAsRefStr, FromString` are already derived through `StringEnum`.

jplatte commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,381 @@
+    RecoveryKey,
+
+    /// Custom secret name.
+    Custom(String),
```
Like in the other similar enum types, the `Custom` variant should be `#[doc(hidden)]` and prefixed with an underscore to make it very clear it's not meant to be used.

Frinksy replied:
The `SecretName` variants aren't mentioned in the "Secrets" module of the unstable spec, but rather in the "End-to-End Encryption" module. That's why I thought that these were only suggestions, and it was expected that any other secret could be stored.
However I see that hiding the `_Custom` variant does not change the functionality, and simplifies the docs for those who only want to implement the "basic" features.

jplatte replied:
Regardless of whether they're suggestions or not, we must not expose `Custom` variants except for some special cases where we know no additional variants are going to ever be introduced, for forwards compatibility. Having `"foo"` deserialize to `SomeEnum::Custom("foo")` first and to `SomeEnum::Foo` later after an update means people's code checking for `SomeEnum::Custom("foo")` will break, without there being any compiler warnings, which is very annoying.

Frinksy committed:
Derive more traits and hide custom variant.

poljar reviewed (approved):
Can confirm, transferring cross signing keys works between the rust-sdk and Element Web. Thanks.

![secret-sharing](https://user-images.githubusercontent.com/552026/127665091-1e57a20c-cee3-4c4c-8368-d28c84e4382f.png)

jplatte commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,381 @@
+    }
+}
+
+impl<'de> Deserialize<'de> for RequestToDeviceEventContent {
```
This impl can be derived; the logic for handling the action and name fields can instead be done in the `Deserialize` impl for `RequestAction` (which can also be implemented with a derive + `#[serde(try_from)]` so it's shorter than that visitor stuff).

Frinksy replied:
I'm not sure if I implemented it the way you meant, but it definitely is shorter and much less complex. Thanks for the pointers!

jplatte commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,381 @@
+    {
+        match self {
+            Self::Request(name) => {
+                let mut st = serializer.serialize_map(Some(2))?;
```
I think you can use `serialize_struct` here, which might allocate a bit less.

jplatte commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,381 @@
+                st.serialize_entry("action", "request_cancellation")?;
+                st.end()
+            }
+            RequestAction::_Custom(custom) => serializer.serialize_str(custom),
```
I don't think this makes sense, wouldn't the custom action still have to be serialized inside an `action` field, same as in the match arm above?

jplatte replied:
Test cases for this would also be useful.

Frinksy replied:
I think you're right, the custom action should still be serialized inside an `action` field. 
However I wonder if the custom variant should contain something else than only a `String`, and something of the form `(String, String)` instead (for the action and name fields).

jplatte replied:
For now this seems good.

jplatte commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,381 @@
+    #[test]
+    fn secret_request_serialization() {
+        let content = RequestToDeviceEventContent::new(
+            RequestAction::Request(SecretName::_Custom("org.example.some.secret".into())),
```
Please don't use private API in tests when it's not necessary. `SecretName` has a `From<&str>` / `From<String>` implementation that can be used here, right?

Frinksy replied:
I removed the uses of `_Custom` where I could, but I think there's no good way around it for the two new tests I added. I don't think it makes sense to have a `From<&str>` impl for `RequestAction`.

jplatte commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,381 @@
+    }
+
+    #[test]
+    fn secret_request_cancellation_deserialisation() {
```
```suggestion
    fn secret_request_cancellation_deserialization() {
```

for consistency with serde and other parts of the code, please use American English.

jplatte commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,381 @@
+            }
+            if secret == "org.example.some.secret"
+            && requesting_device_id == "ABCDEFG"
+            && request_id == "randomly_generated_id_9573"
```
For consistency:

```suggestion
            if secret == "org.example.some.secret"
	              && requesting_device_id == "ABCDEFG"
	              && request_id == "randomly_generated_id_9573"
```

Frinksy committed:
Derive using try_from and cleanup.

jplatte commented:
Looks like CI fails for an unrelated reason, can you try rebasing?

jplatte commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,302 @@
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+#[ruma_event(type = "m.secret.request", kind = ToDevice)]
+pub struct RequestToDeviceEventContent {
+    /// The action for the request, one of `["request", "request_cancellation"]`.
```
```suggestion
    /// The action for the request.
```

The possible values are documented in the `RequestAction` type's docs. Listing them here just increases the likelyhood of these docs becoming out-of-date.

jplatte commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,302 @@
+pub struct RequestToDeviceEventContent {
+    /// The action for the request, one of `["request", "request_cancellation"]`.
+    ///
+    /// If the action is "request", the name of the secret must also be provided.
```
This is enforced by the type, no need to include this note.

jplatte commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,302 @@
+    }
+}
+
+/// Action for a *m.secret.request* event.
```
```suggestion
/// Action for an *m.secret.request* event.
```

jplatte commented on crates/ruma-events/src/secret/request.rs:
```diff
@@ -0,0 +1,302 @@
+}
+
+#[derive(Deserialize)]
+struct RequestActionInit {
```
I'd prefer a different name since our `Init` structs elsewhere are public types with `From`, not `TryFrom` being used to convert them to the "main" type.

```suggestion
struct RequestActionJsonRepr {
```

jplatte reviewed (commented):
Looks good overall, I only have some style- / docs-related concerns.

Frinksy committed:
Cleanup docs and rename RequestActionInit to RequestActionJsonRepr.
